Continual web panic until workers are redeployed

This issue happens rarely, maybe once a month, but the biggest problem is that after it starts it keeps happening across all the web nodes until we resort to redeploying all the workers.

```
panic: runtime error: invalid memory address or nil pointer dereference[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0xb4a25a]goroutine 175554 [running]:github.com/concourse/concourse/atc/exec.(*RetryStep).Succeeded(0xc066d64690, 0x2)
/tmp/build/1c3187db/concourse/atc/exec/retry_step.go:47 +0x2agithub.com/concourse/concourse/atc/exec.OnSuccessStep.Succeeded(0x2841da0, 0xc066d64690, 0x284a7e0, 0xc0680fe400, 0x424101)
/tmp/build/1c3187db/concourse/atc/exec/on_success.go:45 +0x31github.com/concourse/concourse/atc/exec.OnSuccessStep.Succeeded(0x2841e20, 0xc066d647b0, 0x284a7e0, 0xc0680fe4c0, 0xc05ec4cc90)
/tmp/build/1c3187db/concourse/atc/exec/on_success.go:45 +0x5cgithub.com/concourse/concourse/atc/exec.OnFailureStep.Succeeded(...)
/tmp/build/1c3187db/concourse/atc/exec/on_failure.go:44github.com/concourse/concourse/atc/exec.EnsureStep.Succeeded(0x284a7a0, 0xc0680fe640, 0x2841da0, 0xc066d64870, 0xc051486660)
/tmp/build/1c3187db/concourse/atc/exec/ensure_step.go:58 +0x31github.com/concourse/concourse/atc/engine.(*engineBuild).Run(0xc03c79bd60, 0x2895fa0, 0xc00578d020)
/tmp/build/1c3187db/concourse/atc/engine/engine.go:215 +0x8cdcreated by github.com/concourse/concourse/atc/builds.(*Tracker).Track
/tmp/build/1c3187db/concourse/atc/builds/tracker.go:49 +0x425
```

There was a similar issue - https://github.com/concourse/concourse/issues/3989 - so perhaps it was only partially fixed.

## Version Info

* Concourse version: 5.7.0

--- a/atc/configvalidate/validate_test.go
+++ b/atc/configvalidate/validate_test.go
@@ -1504,7 +1504,7 @@ var _ = Describe("ValidateConfig", func() {
 							Step: &atc.PutStep{
 								Name: "some-resource",
 							},
-							Attempts: -1,
+							Attempts: 0,
 						},
 					})
 
@@ -1513,7 +1513,7 @@ var _ = Describe("ValidateConfig", func() {
 
 				It("does return an error", func() {
 					Expect(errorMessages).To(HaveLen(1))
-					Expect(errorMessages[0]).To(ContainSubstring("jobs.some-other-job.plan.do[0].attempts: cannot be negative"))
+					Expect(errorMessages[0]).To(ContainSubstring("jobs.some-other-job.plan.do[0].attempts: must be greater than 0"))
 				})
 			})
 

--- a/atc/exec/retry_step.go
+++ b/atc/exec/retry_step.go
@@ -44,5 +44,8 @@ func (step *RetryStep) Run(ctx context.Context, state RunState) error {
 
 // Succeeded delegates to the last step that it ran.
 func (step *RetryStep) Succeeded() bool {
+	if step.LastAttempt == nil {
+		return false
+	}
 	return step.LastAttempt.Succeeded()
 }

--- a/atc/exec/retry_step_test.go
+++ b/atc/exec/retry_step_test.go
@@ -339,4 +339,16 @@ var _ = Describe("Retry Step", func() {
 			})
 		})
 	})
+
+	Context("when the RetryStep is given no attempts", func() {
+		BeforeEach(func() {
+			step = Retry()
+		})
+
+		Describe("Succeeded", func() {
+			It("should return false", func() {
+				Expect(step.Succeeded()).To(BeFalse())
+			})
+		})
+	})
 })

--- a/atc/exec/retry_step_test.go
+++ b/atc/exec/retry_step_test.go
@@ -40,6 +40,27 @@ var _ = Describe("Retry Step", func() {
 		step = Retry(attempt1, attempt2, attempt3)
 	})
 
+	Context("when calling succeeded before running", func() {
+		Context("when the RetryStep is given no attempts", func() {
+			BeforeEach(func() {
+				step = Retry()
+			})
+
+			Describe("Succeeded", func() {
+				It("should return false", func() {
+					Expect(step.Succeeded()).To(BeFalse())
+				})
+			})
+		})
+
+		Context("when the RetryStep is given attempts", func() {
+			Describe("Succeeded", func() {
+				It("should return false", func() {
+					Expect(step.Succeeded()).To(BeFalse())
+				})
+			})
+		})
+	})
 	Context("when attempt 1 succeeds", func() {
 		BeforeEach(func() {
 			attempt1.SucceededReturns(true)
@@ -336,18 +357,6 @@ var _ = Describe("Retry Step", func() {
 
 					Expect(attempt3.SucceededCallCount()).To(Equal(2))
 				})
-			})
-		})
-	})
-
-	Context("when the RetryStep is given no attempts", func() {
-		BeforeEach(func() {
-			step = Retry()
-		})
-
-		Describe("Succeeded", func() {
-			It("should return false", func() {
-				Expect(step.Succeeded()).To(BeFalse())
 			})
 		})
 	})

--- a/atc/step_validator.go
+++ b/atc/step_validator.go
@@ -264,8 +264,8 @@ func (validator *StepValidator) VisitRetry(step *RetryStep) error {
 	validator.pushContext(".attempts")
 	defer validator.popContext()
 
-	if step.Attempts < 0 {
-		validator.recordError("cannot be negative")
+	if step.Attempts <= 0 {
+		validator.recordError("must be greater than 0")
 	}
 
 	return nil

--- a/release-notes/latest.md
+++ b/release-notes/latest.md
@@ -1,0 +1,3 @@
+#### <sub><sup><a name="5830" href="#5830">:link:</a></sup></sub> fix
+
+* Fix a validation issue where a step can be set with 0 attempts causing the ATC to panic. #5830

